### ISSUE TITLE
How to access 1st argument in function generating 2nd arg candidates?

### ISSUE THREAD

**Issue by sebthom** [chrisant996/clink#101]
I am trying to create completion for a command that uses two arguments `mycommand  [MAIN_TASK] [SUB_TASK]`.

The possible arguments for the second argument depend on the the first argument. How can I access the entered first argument in the function that generates candidates for the second argument?

```lua
local function main_tasks() {
  local candidates = {}
  -- ... here determine main commands by parsing output of "mycommand --help"
  return candidates 
}

local function sub_tasks() {
  local candidates = {}
  local main_task = ... -- <-- how can I get the value of the first argument here?

  -- ... here determine sub commands by parsing  "mycommand main_task --help"

  return candidates
}

clink.argmatcher("mycommand")
  :addarg(main_tasks)
  :addarg(sub_tasks)
```

**Comment by sebthom** [comment]
I already tried to use `line:getcommandoffset()` / `line:getline()`, etc. documented at [clink.html#line](https://chrisant996.github.io/clink/clink.html#line) but this only results in the error `attempt to index global 'line' (a nil value)`

**Comment by sebthom** [comment]
Ok, I figured it out, the function signature needs to be extended with three arguments `word, word_index, line_state`

```lua
local function sub_tasks(word, word_index, line_state) {
  local candidates = {}
  local main_task = line_state:getword(word_index -1)

  -- ... here determine sub commands by parsing  "mycommand main_task --help"

  return candidates
}
```

Maybe you could add this to the documentation at https://chrisant996.github.io/clink/clink.html#functions-as-argument-options. It is not obvious from there that/how the line_state can be accessed. I only found it by "reverse engineering" the clink repo.

**Comment by chrisant996** [comment]
Can you share more context?

From the summary, maybe the section on "Linking Parsers" is what's desired here.

Or maybe a generator.

Using a function there is possible, but it's more limited. It can't do linked parsers and it can't do automatic input line coloring.

**Comment by sebthom** [comment]
As one use case I am using it for docker command completion here https://github.com/sebthom/more-clink-completions/blob/main/src/docker.lua

**Comment by chrisant996** [comment]
Oh, I see now, thanks.

A few thoughts:

1. There isn't currently a way to get "the previous arg" or "arg number X".  There is a way to get words, but not args -- for example "mycommand --flag1 main_task --flag2 /x /y sub_task" might be valid syntax, in which case there isn't a good way to get a specific arg.

2. Argmatchers aren't intended to do any custom parsing.  If custom parsing is desired, then a generator is the expected approach.  Which is more complicated to do, but also more powerful.

3. Clink doesn't have a good way to do what you want right now.  But there is still good news:
    - You can do it up front, but you can't do it deferred.  Instead of using a callback and input line parsing, make an immediate call and use linked argmatchers.  So instead of `:addarg(func_name)` use `:addarg(func_name())`, and then return for example matches like `"main1"..clink.argmatcher():addarg(func_to_get_args("main1"))`.  This solves all the parsing issues and also makes input coloring work.  It has to do all the parsing up front, but it runs only once per session.
    - A deferred approach is necessary in order to scale well as the number of argmatchers increases that invoke exe's to auto-learn args.  This question got me to think through how to add support for argmatchers that want to dynamically parse help text, and I'll add it to the TODO.md file.  It will be easy to add, so probably expect it in the next couple of weeks, depending on available free time.

**Comment by chrisant996** [comment]
Great question and description of the context, thank you.

I've opened issue #102 for an enhancement to support the scenario.

**Comment by sebthom** [comment]
Thanks for your response. I only started coding in Lua two days ago. Argmatchers is the concept that I could grasp from your documentation. Generators looked very complicated and I couldn't wrap my head around it using the docs.
I would prefer seeing Argmatchers getting more flexible instead of having to deal with generators. :-)

Also making flags deferrable would be nice. As they may change based on the command/sub-command.

**Comment by chrisant996** [comment]
> I would prefer seeing Argmatchers getting more flexible instead of having to deal with generators. :-)

Yes, absolutely -- 100% agreement. 😎

> Also making flags deferrable would be nice. As they may change based on the command/sub-command.

Oh, yes, good point, thank you!

**Comment by chrisant996** [comment]
This is turning out to be more complex and will take longer than I thought.  I no long anticipate it coming soon.  It's high up on the wish list, though.

**Comment by chrisant996** [comment]
It turns out using generators and classifiers are much better for this.

Argmatchers have a lot of specialized functionality, and trying to allow conditionally overriding/replacing that functionality on the fly becomes very complicated for your Lua code.  (It's easy to make Clink support it, but it's either very complicated to use, or it's very incomplete and inconsistent functionality.)

Are there specific reasons why generators look complicated?  They're simpler than argmatchers to use, and infinitely flexible.

Here is a sample I put together for how to do exactly what you're asking for.

This could be saved to a file "**command_callbacks.lua**":
```lua
local _priority = 50
local _commands = {}

--------------------------------------------------------------------------------
-- Use `add_command_callbacks` to register `generate` and/or `classify`
-- callbacks for a specific `command`.
--
-- The optional `generate` callback function can generate matches for the
-- current argument in the input line.  The callback takes 2 arguments:
--      `line`              A string with the preceding arguments (no flags).
--      `match_builder`     https://chrisant996.github.io/clink/clink.html#builder
-- The `generate` callback should return a table of strings, and `true` to stop
-- running other generators or `false` or `nil` to continue running other
-- generators.
--
-- The optional `classify` callback function can classify words to apply colors
-- to the input line.  The callback function takes 4 arguments:
--      `line`              A string with the preceding arguments (no flags).
--      `word`              The word to be classified.
--      `word_index`        The index of the word to be classified.  This can be
--                          passed into classifications:classifyword().
--      `classifications`   https://chrisant996.github.io/clink/clink.html#word_classifications
function add_command_callbacks(command, generate, classify)
    _commands[command] = {generate=generate, classify=classify}
end

--------------------------------------------------------------------------------
local function make_parser(line_state)
    if line_state:getwordcount() <= 0 then
        return
    end

    local word_full = line_state:getword(1)
    local word_name = path.getname(word_full)

    local found = _commands[word_full] or _commands[word_name]
    if not found then
        return
    end

    local parser = {}
    parser.index = 0
    parser.line_state = line_state
    parser.generate = found.generate
    parser.classify = found.classify

    function parser:advance()
        if self:atend() then
            return false
        end
        self.index = self.index + 1
        return true
    end

    function parser:getword()
        return self.line_state:getword(self.index)
    end

    function parser:isflag()
        if self:atend() then
            return false
        end
        return self:getword():sub(1,1) == '-'
    end

    function parser:atend()
        if self.index + 1 > self.line_state:getwordcount() then
            return true
        end
    end

    return parser
end

--------------------------------------------------------------------------------
local generator = clink.generator(priority)
function generator:generate(line_state, match_builder)
    local parser = make_parser(line_state)
    if parser and parser.generate and parser:advance() then
        local interrogate = ""
        repeat
            local word = parser:getword()
            if not parser:isflag() then
                if #interrogate > 0 and #word > 0 then
                    interrogate = interrogate.." "
                end
                interrogate = interrogate..word
            end
        until not parser:advance()
        if #interrogate > 0 then
            return parser.generate(interrogate, match_builder)
        end
    end
end

--------------------------------------------------------------------------------
local classifier = clink.classifier(priority)
function classifier:classify(commands)
    for _,command in ipairs(commands) do
        local parser = make_parser(command.line_state)
        if parser and parser.classify and parser:advance() then
            local interrogate = ""
            repeat
                local word = parser:getword()
                parser.classify(interrogate, word, parser.index, command.classifications)
                if not parser:isflag() then
                    if #interrogate > 0 then
                        interrogate = interrogate.." "
                    end
                    interrogate = interrogate..word
                end
            until not parser:advance()
        end
    end
end
```

And here is an example "mycommand" using the above.
This could be saved as "**mycommand.lua**":

```lua
--------------------------------------------------------------------------------
-- Sample code for adding `mycommand`:

local mycommand_args = { "aaa", "bbb" }
local mycommand_aaa_args = { "111", "222", "333" }
local mycommand_bbb_args = { "444", "555", "666" }

local function mycommand_generate(line, match_builder)
    if line == "mycommand" then
        -- ... parse arguments from `mycommand --help` ...
        match_builder:addmatches(mycommand_args)
    elseif line == "mycommand aaa" then
        -- ... parse arguments from `mycommand aaa --help` ...
        match_builder:addmatches(mycommand_aaa_args)
    elseif line == "mycommand bbb" then
        -- ... parse arguments from `mycommand bbb --help` ...
        match_builder:addmatches(mycommand_bbb_args)
    end

    return true
end

local function mycommand_classify(line, word, word_index, classifications)
    if line == "" then
        if word == "mycommand" then
            classifications:classifyword(word_index, "m")
        end
        return
    end

    local args = nil
    if line == "mycommand" then
        args = mycommand_args
    elseif line == "mycommand aaa" then
        args = mycommand_aaa_args
    elseif line == "mycommand bbb" then
        args = mycommand_bbb_args
    end

    if args then
        for _,arg in ipairs(args) do
            if arg == word then
                classifications:classifyword(word_index, "a")
                break
            end
        end
    end
end

add_command_callbacks("mycommand", mycommand_generate, mycommand_classify)
```

**Comment by chrisant996** [comment]
I ran into a similar case:

Premake5 loads `premake.lua` scripts based on the current directory, and the scripts define the available commands.

So, it isn't possible to make a pre-defined argmatcher for Premake5, because the available actions and flags are inherently dynamic.

That has renewed my interest in exploring this issue further.

A solution would ideally:
- Allow an argmatcher to dynamically extend or reset itself.
- Support match generation.
- Support input line coloring.
  - Use a coroutine so that extending the argmatcher (e.g. by running `some_program --help` and parsing the output) can extend the argmatcher dynamically and update the input line coloring, without making typing non-responsive.

I am playing with some potential ways to accommodate that...

**Comment by chrisant996** [comment]
**UPDATE:**  Clink v1.3.10 will introduce an easy-to-use way to initialize (or re-initialize) an argmatcher on the fly.

Essentially it's this:

1. Create an argmatcher.
2. Set a callback function for (re)initializing it on demand (i.e. defer initialization until the argmatcher is actually used).
    - The callback function decides whether (re)initialization is needed (e.g. Premake5 actions and flags need to be re-initialized when the current directory changes).
    - The callback function can reset the argmatcher and initialize the argmatcher.

For example, the callback function could use `io.popen("premake5 --help 2>nul")` and parse the help text, and use `:addflags()` and `:addarg()` to initialize the argmatcher with the appropriate flags and arguments.

Note: The callback function is always executed as a coroutine. This means any `os.execute()` and `io.popen()` calls automatically allow typing input to remain responsive.

**Comment by chrisant996** [comment]
v1.3.10 is published, and the `premake_argmatcher.lua` script is in the Discussions section.
